API Libraries > Polkadot API (PAPI) (#138)
polkadot-developers/polkadot-docs: develop/toolkit/api-libraries/papi.md

## Get Started

### API Instantiation

To instantiate the API, you can install the package by using the following command:

=== "npm"
    ```bash
    npm i polkadot-api
    ```

=== "pnpm"
    ```bash
    pnpm add polkadot-api
    ```

=== "yarn"
    ```bash
    yarn add polkadot-api
    ```

Then, obtain the latest metadata from the target chain and generate the necessary types:

```bash
# Add the target chain
npx papi add dot -n polkadot
```

> The `papi add` command initializes papi add, assigns the chain a custom name and specifies downloading metadata from the Polkadot chain. You can replace `dot` with the chain name if you want to add a different chain. Once the latest metadata is downloaded, generate the required types:

Author: You are right; it's confusing. I paraphrased it in [20bc1f5](https://github.com/polkadot-developers/polkadot-docs/pull/138/commits/20bc1f55e7174a1d9a4ebb54b329389955d05571)

```bash
# Generate the necessary types
npx papi
```

You can now set up a [`PolkadotClient`](https://github.com/polkadot-api/polkadot-api/blob/main/packages/client/src/types.ts#L153){target=\_blank} with your chosen provider to begin interacting with the API. Choose from Smoldot via WebWorker, Node.js, or direct usage, or connect through the WSS provider. The examples below show how to configure each option for your setup.

=== "Smoldot(WebWorker)"

    ```typescript
    --8<-- "code/develop/toolkit/api-libraries/papi/api-instantiation-smoldot-webworker.ts"
    ```

=== "Smoldot(Node.js)"

    ```typescript
    --8<-- "code/develop/toolkit/api-libraries/papi/api-instantiation-smoldot-nodejs.ts"
    ```

=== "Smoldot"

    ```typescript
    --8<-- "code/develop/toolkit/api-libraries/papi/api-instantiation-smoldot.ts"
    ```

=== "WSS"

    ```typescript
    --8<-- "code/develop/toolkit/api-libraries/papi/api-instantiation-wss.ts"
    ```

Now that you have set up the client, you can interact with the chain by reading and sending transactions.

### Reading Chain Data

With the `TypedApi`, you can read data from the blockchain using the following patterns:

> - To read a **constant** of the blockchain, you can use the `constants` interface and access the constant you need:

Author: it was just to emphasize the "type" of query or request we are doing to the chain, like you can either query a `constant`, `storage` portions of the chain OR do queries via `runtime apis`

<img width="864" alt="image" src="https://github.com/user-attachments/assets/1c79f213-58d6-401c-a665-82d2325bfe41">

@eshaben, do you believe there is another way to be more explicit in achieving the same? I'm not 100% convinced with this approach, but it's the only one that came to my mind tbh

> - To read a **constant** of the blockchain, you can use the `constants` interface and access the constant you need:

Reviewer: Maybe start the list item with each one:

```
The `TypedApi` provides a streamlined way to read blockchain data through three main interfaces, each designed for specific data access patterns:

- **Constants** - access fixed values or configurations on the blockchain using the `constants` interface:

    ```typescript
    const version = await typedApi.constants.System.Version();
    ```

- **Storage queries** - retrieve stored values by querying the blockchain’s storage via the `query` interface:

    ```typescript
    const asset = await api.query.ForeignAssets.Asset.getValue(
      token.location,
      { at: 'best' },
    );
    ```

- **Runtime APIs** - interact directly with runtime APIs using the `apis` interface:

    ```typescript
    const metadata = await typedApi.apis.Metadata.metadata();
    ```
```

> - To read a **constant** of the blockchain, you can use the `constants` interface and access the constant you need:

Author: makes sense, it's much clearer. Ty!

    ```typescript
    const version = await typedApi.constants.System.Version();
    ```

- To read a value from the **storage**, you can use the `query` interface with the following pattern:

    ```typescript
    const asset = await api.query.ForeignAssets.Asset.getValue(
      token.location,
      { at: 'best' },
    );
    ```

- To directly query a **Runtime API**, you can use the `apis` interface and access the API you need, for example:

    ```typescript
    const metadata = await typedApi.apis.Metadata.metadata();
    ```

> To learn more about the different actions you can perform with the `TypedApi`, refer to the [Typed API Reference](https://papi.how/typed){target=\_blank}.

Reviewer: ```suggestion
To learn more about the different actions you can perform with the `TypedApi`, refer to the [TypedApi reference](https://papi.how/typed){target=\_blank}.
```

    ``` 

    For instance, to execute the `balances.transferKeepAlive` call, you can use the following snippet:

    ```typescript
    --8<-- "code/develop/toolkit/api-libraries/papi/send-tx-transfer-keep-alive.ts"
    ```

    Ensure you replace `INSERT_DESTINATION_ADDRESS` and `INSERT_VALUE` with the actual destination address and value, respectively.

- The `txFromCallData` method allows you to send a transaction using the call data. This option accepts binary call data and constructs the transaction from it. It validates the input upon creation and will throw an error if invalid data is provided. The pattern is as follows:

    ```typescript
    const callData = Binary.fromHex('0x...');
    const tx: Transaction = typedApi.txFromCallData(callData);
    ``` 

    For instance, to execute a transaction using the call data, you can use the following snippet:

    ```typescript
    const callData = Binary.fromHex('0x00002470617065726d6f6f6e');
    const tx: Transaction = typedApi.txFromCallData(callData);
    ```

For more information about sending transactions, refer to the [Transactions](https://papi.how/typed/tx#transactions){target=\_blank} page.

## Where to Go Next

> For an in-depth guide on how to use PAPI, refer to the official [PAPI](https://papi.how/typed){target=\_blank} documentation.

Reviewer: ```suggestion
For an in-depth guide on how to use PAPI, refer to the official [PAPI](https://papi.how/){target=\_blank} documentation.
```

```bash
# Generate the necessary types
npx papi
```

You can now set up a [`PolkadotClient`](https://github.com/polkadot-api/polkadot-api/blob/main/packages/client/src/types.ts#L153){target=\_blank} with your chosen provider to begin interacting with the API. Choose from Smoldot via WebWorker, Node.js, or direct usage, or connect through the WSS provider. The examples below show how to configure each option for your setup.

=== "Smoldot(WebWorker)"

    ```typescript
    --8<-- "code/develop/toolkit/api-libraries/papi/api-instantiation-smoldot-webworker.ts"
    ```

=== "Smoldot(Node.js)"

    ```typescript
    --8<-- "code/develop/toolkit/api-libraries/papi/api-instantiation-smoldot-nodejs.ts"
    ```

=== "Smoldot"

    ```typescript
    --8<-- "code/develop/toolkit/api-libraries/papi/api-instantiation-smoldot.ts"
    ```

=== "WSS"

    ```typescript
    --8<-- "code/develop/toolkit/api-libraries/papi/api-instantiation-wss.ts"
    ```

Now that you have set up the client, you can interact with the chain by reading and sending transactions.

### Reading Chain Data

With the `TypedApi`, you can read data from the blockchain using the following patterns:

> - To read a **constant** of the blockchain, you can use the `constants` property and access the constant you need:

Reviewer 2: ```suggestion
- To read a **constant** of the blockchain, you can use the `constants` interface and access the constant you need:
```

    ```typescript
    const version = await typedApi.constants.System.Version()
    ```

> - To read a value from the **storage**, you can use the following pattern:

Reviewer 2: ```suggestion
- To read a value from the **storage**, you can use the `query` interface with the following pattern:
```

    ```typescript
    const result = await typedApi.query.Pallet.Query.getValue(
        arg1,
        arg2,
        arg3,
        { at: "best" }

>     )

Reviewer 2: The example could be clearer. Replace with a specific storage read.

    ```

> - To directly query a **Runtime API**, you can use the `query` property and access the API you need, for example:

Reviewer 2: ```suggestion
- To directly query a **Runtime API**, you can use the `apis` interface and access the API you need, for example:
```

    ```typescript
    const tx: Transaction = typedApi.tx.Pallet.Call({arg1, arg2, arg3})
    ``` 

    For instance, to execute the `balances.transferKeepAlive` call, you can use the following snippet:

    ```typescript
    --8<-- "code/develop/toolkit/api-libraries/papi/send-tx-transfer-keep-alive.ts"
    ```

    Ensure you replace `INSERT_DESTINATION_ADDRESS` and `INSERT_VALUE` with the actual destination address and value, respectively.

- The `txFromCallData` method allows you to send a transaction using the call data. This option accepts binary call data and constructs the transaction from it. It validates the input upon creation and will throw an error if invalid data is provided. The pattern is as follows:

    ```typescript
    const callData = Binary.fromHex("0x...")
    const tx: Transaction = typedApi.txFromCallData(callData)
    ``` 

    For instance, to execute a transaction using the call data, you can use the following snippet:

    ```typescript
    const callData = Binary.fromHex("0x00002470617065726d6f6f6e")
    const tx: Transaction = typedApi.txFromCallData(callData)
    ```

>     Ensure you replace `INSERT_CALL_DATA` with the actual call data, a hex string.

Reviewer 2: There is no INSERT_CALL_DATA in the snippet
